core/controllers/voice_artist.py

```diff
@@ -41,69 +40,63 @@ class AudioUploadHandler(
     to the local datastore in dev).
     """
 
+    GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
+
+    URL_PATH_ARGS_SCHEMAS = {
+        'exploration_id': {
+            'schema': {
+                'type': 'basestring',
+                'validators': [{
+                    'id': 'is_regex_matched',
+                    'regex_pattern': constants.ENTITY_ID_REGEX
+                }]
+            }
+        }
+    }
+
+    HANDLER_ARGS_SCHEMAS = {
+        'POST': {
+            'raw_audio_file': {
+                'schema': {
+                    'type': 'basestring',
+                    'validators': [{
+                        'id': 'is_valid_audio_file'
+                    }]
+                }
+            },
+            'filename': {
+                'schema': {
+                    'type': 'basestring',
+                    'validators': [{
+                        'id': 'is_regex_matched',
+                        'regex_pattern': r'[^\s]+(\.(?i)(mp3))$'
+                    }]
+                }
+            }
+        }
+    }
+
     # The string to prefix to the filename (before tacking the whole thing on
     # to the end of 'assets/').
     _FILENAME_PREFIX = 'audio'
 
     @acl_decorators.can_voiceover_exploration
     def post(self, exploration_id: str) -> None:
         """Saves an audio file uploaded by a content creator."""
-        raw_audio_file = self.request.get('raw_audio_file')
-        filename = self.payload.get('filename')
-        allowed_formats = list(feconf.ACCEPTED_AUDIO_EXTENSIONS.keys())
-
-        if not raw_audio_file:
-            raise self.InvalidInputException('No audio supplied')
-        dot_index = filename.rfind('.')
-        extension = filename[dot_index + 1:].lower()
-
-        if dot_index in (-1, 0):
-            raise self.InvalidInputException(
-                'No filename extension: it should have '
-                'one of the following extensions: %s' % allowed_formats)
-        if extension not in feconf.ACCEPTED_AUDIO_EXTENSIONS:
-            raise self.InvalidInputException(
-                'Invalid filename extension: it should have '
-                'one of the following extensions: %s' % allowed_formats)
+        assert self.normalized_payload is not None
+        assert self.normalized_request is not None
+
+        filename = self.normalized_payload['filename']
+        # Here we use cast because we want to narrow down the type
+        # of normalized_request['raw_audio_file'] from  str to bytes.
+        raw_audio_file = cast(bytes, self.normalized_request['raw_audio_file'])
```

Comment:
```suggestion
        raw_audio_file = self.normalized_request['raw_audio_file'].encode('ascii')
```

Reply:
I tried this initially, but doing this returns the same error for several tests: 

`{"error": "'bytes' object has no attribute 'encode'", "status_code": 500}`
caused by
```
File "~/oppia/core/controllers/voice_artist.py", line 92, in post
    raw_audio_file = self.normalized_request['raw_audio_file'].encode('ascii')
AttributeError: 'bytes' object has no attribute 'encode'
```

I think this is because the `raw_audio_file` is sent as bytes, even though the type hint for `normalized_request` is `Dict[str,str]`. In the schema for `raw_audio_file`, the type is `basestring` which means both `str` and `bytes` should be okay, so I'm not sure what the best solution for this would be?

Furthermore, only 3 tests fails with this change. I think tests that should expect a 200 OK (but don't currently), such as https://github.com/oppia/oppia/blob/28c396663b6616e38e42d558d829dad897064f68/core/controllers/resources_test.py#L579-L595 should fail. I'll make these changes as well to the tests.

Reply:
Ah I know what is going on, you will need to add the correct typing. @sahiljoster32 Can you help here?

Reply:
Hi @adamhalim,

From the discussion, it looks to me that `self.normalized_request['raw_audio_file']` should be typed as bytes rather than str.

So, I think you should define a typedDict for`self.normalized_request`. As we are doing in other files also https://github.com/oppia/oppia/blob/e915b3dccfcc629800b6fa99ccc2817e4d7ae562/core/controllers/editor.py#L94

Reply:
Ok, for your case it goes something like below:

```python
class AudioUploadHandlerNormalizedRequestDict(TypedDict):
    """Dict representation of AudioUploadHandler's
    normalized_request dictionary.
    """

    raw_audio_file: bytes


```

and pass that `AudioUploadHandlerNormalizedRequestDict` type as a second type argument in basehandler `BaseHandler[Dict[str, str], AudioUploadHandlerNormalizedRequestDict] ` like below:
```python
class AudioUploadHandler(
    base.BaseHandler[Dict[str, str], AudioUploadHandlerNormalizedRequestDict]
):
```

Reply:
Hey @sahiljoster32 , thanks for your comments.

I just tried to do this and got the following Mypy type check error: `core/controllers/voice_artist.py:101: error: Redundant cast to "AudioUploadHandlerNormalizedRequestDict"  [redundant-cast]`

These are the changes I did:
Added RequestDict:
```python

class AudioUploadHandlerNormalizedRequestDict(TypedDict):
    """Dict representation of AudioUploadHandler's
    normalized_request dictionary.
    """

    raw_audio_file: bytes
```
Changed handler arguments:
```python
class AudioUploadHandler(
    base.BaseHandler[Dict[str, str], AudioUploadHandlerNormalizedRequestDict]
):
```

Cast in post method:
```python
        # Here we use cast because we want to narrow down the type
        # of 'normalized_request' to a particular TypedDict that
        # was defined according to the schemas so that the type
        # of fetched value is considered bytes and not str.
        request_data = cast(
            AudioUploadHandlerNormalizedRequestDict,
            self.normalized_request
        )
        raw_audio_file = request_data['raw_audio_file']
        filename = self.normalized_payload['filename']

```

Reply:
explicit Cast in post method is not needed.

```python
        # Here we use cast because we want to narrow down the type
        # of 'normalized_request' to a particular TypedDict that
        # was defined according to the schemas so that the type
        # of fetched value is considered bytes and not str.
        request_data = cast(
            AudioUploadHandlerNormalizedRequestDict,
            self.normalized_request
        )
```
Now, you can just simply use 
```python
raw_audio_file = self.normalized_request['raw_audio_file']
```

Reply:
The other two changes are good!

Reply:
Awesome, thanks! Just pushed the new changes